controllers/glanceapi_controller.go

```diff
@@ -1427,3 +1486,94 @@ func (r *GlanceAPIReconciler) glanceAPIRefresh(
 	}
 	return nil
 }
+
+// ensureGlanceAPITopology - retrieve the associated Topology/Affinity CR
+func (r *GlanceAPIReconciler) ensureGlanceAPITopology(
+	ctx context.Context,
+	h *helper.Helper,
+	instance *glancev1.GlanceAPI,
+) (*topologyv1.Topology, error) {
+
+	var err error
+	var hash string
+	ns := instance.Namespace
+
+	if instance.Spec.Topology.Namespace != "" {
+		ns = instance.Spec.Topology.Namespace
+	}
+	topology, hash, err := topologyv1.GetTopologyByName(
+		ctx,
+		h,
+		instance.Spec.Topology.Name,
+		ns,
+	)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			return topology, err
+		}
+		return topology, err
+	}
+
+	// Add finalizer to the resource because it is going to be consumed by GlanceAPI
+	if !controllerutil.ContainsFinalizer(topology, h.GetFinalizer()) {
+		controllerutil.AddFinalizer(topology, fmt.Sprintf("%s-%s", h.GetFinalizer(), instance.APIName()))
+	}
+	// Update the resource
+	if err := h.GetClient().Update(ctx, topology); err != nil {
+		return topology, err
+	}
+	if hashMap, changed := util.SetHash(instance.Status.Hash, "topology", hash); changed {
+		instance.Status.Hash = hashMap
+		r.Log.Info(fmt.Sprintf("Set Topology hash %s - %s", "topology", hash))
+	}
+	return topology, nil
+}
+
+// ensureDeletedTopology -
+func (r *GlanceAPIReconciler) ensureDeletedTopology(
+	ctx context.Context,
+	instance *glancev1.GlanceAPI,
+	h *helper.Helper,
+) (ctrl.Result, error) {
+	ns := instance.Namespace
+
+	// no Topology is passed to the GlanceAPI, and it was not used before
+	if instance.Status.LastAppliedTopology == "" {
+		return ctrl.Result{}, nil
+	}
+
+	// Topology is referenced in the .Spec, check the namespace
+	if instance.Spec.Topology != nil {
+		// Check namespace
+		if instance.Spec.Topology.Namespace != "" {
+			ns = instance.Spec.Topology.Namespace
+		}
+	}
+
+	name := instance.Status.LastAppliedTopology
+
+	// Remove the finalizer from the Topology CR
+	topology, _, err := topologyv1.GetTopologyByName(
+		ctx,
+		h,
+		name,
+		ns,
+	)
+
+	if k8s_errors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	}
+	if err != nil && !k8s_errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+	if err == nil {
+		if controllerutil.RemoveFinalizer(topology, fmt.Sprintf("%s-%s", h.GetFinalizer(), instance.APIName())) {
+			err = r.Update(ctx, topology)
+			if err != nil && !k8s_errors.IsNotFound(err) {
+				return ctrl.Result{}, err
+			}
+			util.LogForObject(h, "Removed finalizer from Topology", topology)
+		}
+	}
+	return ctrl.Result{}, err
```

Comment:
should be the same like this, right?


```
        if err != nil && !k8s_errors.IsNotFound(err) {
		return ctrl.Result{}, err
	}

	if !k8s_errors.IsNotFound(err){
		if controllerutil.RemoveFinalizer(topology, fmt.Sprintf("%s-%s", h.GetFinalizer(), instance.APIName())) {
			err = r.Update(ctx, topology)
			if err != nil && !k8s_errors.IsNotFound(err) {
				return ctrl.Result{}, err
			}
			util.LogForObject(h, "Removed finalizer from Topology", topology)
		}
	}
	return ctrl.Result{}, nil
```

Reply:
good catch I guess is the same yes. Updating the code.

Reply:
Done